### ISSUE TITLE
regular expression bug (design limitation?)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6504]
Migrated from [rt.perl.org#22182](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22182) (status was 'resolved')

Searchable as RT22182$

**Comment by p5pRT** [comment]
### From esp5@mdss03.comp.pge.com

#### Created by esp5@mdss03.comp.pge.com



The following script dumps core on both perl\-5\.8\.0 and perl\-5\.6\.1 \- it's probably
due to a design limitation in the regular expression engine\, and I'm interested to
hear any workarounds\.


\-\-\-
script
\-\-\-
$\_singlestring = q$\(?>\\'\(?>\[^\\\\\\'\]\+|\\\\\\\.\)\*\\'\)$;
my $\_parens = "\(?>\\\\s\*\\\\\(\(?>$\_singlestring|\(?>\[^'\\\\\(\\\\\)\]\*\)\)\+\\\\\)\)";

my $a = '\(' \. "'asdfasdf'" x 300000 \. '\)';

if \($a =~ m"\($\_parens\)"\) \{ print STDERR "HERE"; \}
\-\-\-


Before dying\, it gets the following trace&#8203;:

\#0  breakit \(\) at regexec\.c&#8203;:4330
\#1  0x000dcb1c in S\_reginclass \(n=0x16cc20\, 
&nbsp;    p=0xf66064 "''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf"\.\.\.\, lenp=0x0\, do\_utf8=0 '\\0'\) at regexec\.c&#8203;:4338
\#2  0x000db984 in S\_regrepeat \(p=0x16cc20\, max=2147483647\) at regexec\.c&#8203;:4052
\#3  0x000da87c in S\_regmatch \(prog=0x16cc20\) at regexec\.c&#8203;:3707
\#4  0x000db174 in S\_regmatch \(prog=0x16cc10\) at regexec\.c&#8203;:3847
\#5  0x000d8af8 in S\_regmatch \(prog=0x16cc64\) at regexec\.c&#8203;:3214
\#6  0x000d8af8 in S\_regmatch \(prog=0x16cc64\) at regexec\.c&#8203;:3214
\#7  0x000d8434 in S\_regmatch \(prog=0x16cc08\) at regexec\.c&#8203;:3049
\#8  0x000db174 in S\_regmatch \(prog=0x16cbf8\) at regexec\.c&#8203;:3847
\#9  0x000db174 in S\_regmatch \(prog=0x16cbec\) at regexec\.c&#8203;:3847
\#10 0x000d8af8 in S\_regmatch \(prog=0x16ccc8\) at regexec\.c&#8203;:3214
\#11 0x000d8af8 in S\_regmatch \(prog=0x16ccc8\) at regexec\.c&#8203;:3214
\#12 0x000d8af8 in S\_regmatch \(prog=0x16ccc8\) at regexec\.c&#8203;:3214
\#13 0x000d8af8 in S\_regmatch \(prog=0x16ccc8\) at regexec\.c&#8203;:3214
\#14 0x000d8af8 in S\_regmatch \(prog=0x16ccc8\) at regexec\.c&#8203;:3214
\#15 0x000d8af8 in S\_regmatch \(prog=0x16ccc8\) at regexec\.c&#8203;:3214
\#16 0x000d8af8 in S\_regmatch \(prog=0x16ccc8\) at regexec\.c&#8203;:3214
\.\.\.
\#17570 0x000d8af8 in S\_regmatch \(prog=0x16ccc8\) at regexec\.c&#8203;:3214
\#17571 0x000d5fa0 in S\_regtry \(prog=0x16cb80\,
&nbsp;    startpos=0xf3b288 "\('asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf"\.\.\.\) at regexec\.c&#8203;:2145
&nbsp;    \#17572 0x000d59dc in Perl\_regexec\_flags \(prog=0x16cb80\,
&nbsp;        stringarg=0xf3b288 "\('asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf"\.\.\.\, strend=0x1217949 "\)"\,
&nbsp;            strbeg=0xf3b288 "\('asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf''asdfasdf"\.\.\.\, minend=15970952\, sv=0x163ed4\, data=0x0\,
&nbsp;    flags=3\) at regexec\.c&#8203;:1973
\#17573 0x000841f4 in Perl\_pp\_match \(\) at pp\_hot\.c&#8203;:1291
\#17574 0x0006dfbc in Perl\_runops\_debug \(\) at dump\.c&#8203;:1398
\#17575 0x00027ad4 in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:1681

which seems to indicate to me that each time a $\_singlestring is matched\, perl calls
S\_regmatch recursively\. After too many S\_regmatch stack levels happen\, gcc gets confused 
and it dumps core\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.0:

Configured by esp5 at Mon May 12 15:29:52 PDT 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.8, archname=sun4-solaris
    uname='sunos mdss03 5.8 generic_108528-15 sun4u sparc sunw,sun-fire-880 '
    config_args='-ds -e -Dprefix=/export/home/esp5/install/debug -Dcc=/home/esp5/install/bin/gcc -Dccflags= -g'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='/home/esp5/install/bin/gcc', ccflags ='-g -g -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DLEAKTEST -DDEBUGGING',
    optimize='-O',
    cppflags='-g -fno-strict-aliasing'
    ccversion='', gccversion='3.2.2', gccosandvers='solaris2.8'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='/home/esp5/install/bin/gcc', ldflags =' -L/usr/local/lib '
    libpth=/usr/local/lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldbm -ldl -lm -lc
    perllibs=-lsocket -lnsl -ldl -lm -lc
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /export/home/esp5/install/debug/lib/perl5/5.8.0/sun4-solaris
    /export/home/esp5/install/debug/lib/perl5/5.8.0
    /export/home/esp5/install/debug/lib/perl5/site_perl/5.8.0/sun4-solaris
    /export/home/esp5/install/debug/lib/perl5/site_perl/5.8.0
    /export/home/esp5/install/debug/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/esp5
    LANG (unset)
    LANGUAGE (unset)
    LC_COLLATE=en_US.ISO8859-1
    LC_CTYPE=en_US.ISO8859-1
    LC_MESSAGES=C
    LC_MONETARY=en_US.ISO8859-1
    LC_NUMERIC=en_US.ISO8859-1
    LC_TIME=en_US.ISO8859-1
    LD_LIBRARY_PATH=/opt/work/install/lib:/home/esp5/install/lib
    LOGDIR (unset)
    PATH=/home/esp5/install/bin:/opt/work/install/bin:/opt/oracle/817/bin:/opt/install/bin:/usr/ccs/bin:/usr/local/bin:/opt/SUNWspro/bin:/bin:/usr/bin:/usr/ucb:/etc:.:/usr/sbin:/usr/etc:/usr/openwin/bin:/usr/lib/lp/postscript:/export/home/dehc/x3270-3.1.1
    PERL_BADLANG (unset)
    PERL_PREFIX=/home/esp5/install
    SHELL=tcsh

```
</details>

**Comment by p5pRT** [comment]
### From sky@nanisky.com


On tisdag\, maj 13\, 2003\, at 01&#8203;:35 Europe/Stockholm\, 
esp5@&#8203;mdss03\.comp\.pge\.com \(via RT\) wrote&#8203;:

>
> The following script dumps core on both perl\-5\.8\.0 and perl\-5\.6\.1 \- 
> it's probably
> due to a design limitation in the regular expression engine\, and I'm 
> interested to
> hear any workarounds\.
>
>
>

There regex engine should be rewritten to be iterative instead of 
recursive\, authority has said that this should be simple but noone has 
stepped up with patches\.


Arthur

**Comment by p5pRT** [comment]
### From ilya@Math.Berkeley.EDU

On Wed\, May 14\, 2003 at 03&#8203;:07&#8203;:05PM \-0700\, Edward Peschko wrote&#8203;:

Sorry\, I could not guess what you have in mind\.\.\.

Yours\,
Ilya

> I'd like to be able to define a coderef that looks like the following&#8203;:
> 
> my $\_coderef = sub \{ my \($text\) = @&#8203;\_;
>                     \# text starts from position at which regular expression last 
>                     \# matched\.
>                     \# do something with $text\, including regular expressions to modify
>                     \# pos\($text\);
>                     \# change pos so that the regular expression 'picks up' where the 
>                     \# sub left off \(ie&#8203;: pos\)\.
>             \}
> 
> Then people could write stuff to do particularly difficult regular expressions using 
> Inline&#8203;::C\, and you could say&#8203;:
> 
> if \($varb =~ m"\($balanced\_parens\)"\)  \# is a C call\.
> \{
> 
> \}
> 
> which IMO is far more intuitive \(and cleaner\) than anything to do with \(?\.\.\.\)\.
> 
> Ed

**Comment by p5pRT** [comment]
### From esp5@mdssirds.comp.pge.com

On Tue\, May 13\, 2003 at 03&#8203;:36&#8203;:07PM \-0000\, A\.Bergman \(via RT\) wrote&#8203;:
> 
> On tisdag\, maj 13\, 2003\, at 01&#8203;:35 Europe/Stockholm\, 
> esp5@&#8203;mdss03\.comp\.pge\.com \(via RT\) wrote&#8203;:
> 
> >
> > The following script dumps core on both perl\-5\.8\.0 and perl\-5\.6\.1 \- 
> > it's probably
> > due to a design limitation in the regular expression engine\, and I'm 
> > interested to
> > hear any workarounds\.
> >
> >
> >
> 
> There regex engine should be rewritten to be iterative instead of 
> recursive\, authority has said that this should be simple but noone has 
> stepped up with patches\.

I was thinking about this a bit\, \(especially in conjunction with 'push\(@&#8203;INC\, \\&ref\)\)\, 
and was wondering how easy/acceptable it would be to extend the regex engine \(yet again\) 
\- in a way that would be relatively painless\.\.

I'd like to be able to define a coderef that looks like the following&#8203;:

my $\_coderef = sub \{ my \($text\) = @&#8203;\_;
&nbsp;                    \# text starts from position at which regular expression last 
&nbsp;                    \# matched\.
&nbsp;                    \# do something with $text\, including regular expressions to modify
&nbsp;                    \# pos\($text\);
&nbsp;                    \# change pos so that the regular expression 'picks up' where the 
&nbsp;                    \# sub left off \(ie&#8203;: pos\)\.
&nbsp;            \}

Then people could write stuff to do particularly difficult regular expressions using 
Inline&#8203;::C\, and you could say&#8203;:

if \($varb =~ m"\($balanced\_parens\)"\)  \# is a C call\.
\{

\}

which IMO is far more intuitive \(and cleaner\) than anything to do with \(?\.\.\.\)\.

Ed

**Comment by p5pRT** [comment]
### From esp5@mdssirds.comp.pge.com

On Wed\, May 14\, 2003 at 10&#8203;:47&#8203;:06PM \-0000\, Ilya Zakharevich wrote&#8203;:
> On Wed\, May 14\, 2003 at 03&#8203;:07&#8203;:05PM \-0700\, Edward Peschko wrote&#8203;:
> 
> Sorry\, I could not guess what you have in mind\.\.\.
> 
> Yours\,

Suppose you had the following code&#8203;:

my $string = 'abcdefg';

sub \_position\_increment 
\{ 
&nbsp;    my \($string\,$pos\) = @&#8203;\_;
&nbsp;    return\($pos\+3\);
\}

my $\_coderef = \\&\_position\_increment;

if \($string =~ m"\($\_coderef\)"\)
\{
&nbsp;    print $1;   \# prints 'abc';
\}

I'd want \- the moment the regular expression sees $\_coderef\, for perl to run the function
underlying coderef\, passing it the string and the position at which the string was at
when coderef was run\.

So \- stepping through&#8203;:

my $string = 'abcdefg'; \# pos\($string\) == 0;

sub \_position\_increment 
\{ 
&nbsp;    my \($string\, $pos\) = @&#8203;\_;            \# when we see this function\, increment pos by 3
&nbsp;    return\($pos\+3\);                     \# return the new position because we are 
\}                                       \# automatically succesful\. \(if we wanted a non
&nbsp;                                        \# successful match\, return\(\-1\);
&nbsp;                                        
&nbsp;                                        
&nbsp;                                        
my $\_coderef = \\&\_position\_increment;
if \($string =~ m"\($\_coderef\)"\)
\{
&nbsp;    print STDERR "$1";                  \# the engine knows we went in with a pos of 0
&nbsp;                                        \# and came out with a pos of 3\, hence it saves
&nbsp;                                        \# 'abc' in backreference
\}

There's a hook into @&#8203;INC to do a similar thing&#8203;:

push\(@&#8203;INC\, \\&function\_modifying\_inc\_at\_point\_of\_use\);

which runs a coderef each time an '@&#8203;INC' is seen\.

Anyways\, the point is that the proposed syntax is a lot more intuitive than \(?\.\.\.\) since 
the syntax is cleaner\, its backwards compatible with the current regex engine\, and it 
integrates better with C code\. Ultimately\, I'd like to be able to write&#8203;:

int \_parens\( char \*str\, int pos\)
\{
&nbsp;    int len = strlen\(str\);
&nbsp;    while \(str\[pos\+\+\] \!= '\(' && pos \< len\) \{ \}
&nbsp;    while \(str\[pos\+\+\] \!= '\)' && pos \< len\) \{ \}
&nbsp;            
&nbsp;    if \(pos == len\) return\(\-1\)
&nbsp;    return\(pos\);
\}

my $\_parens = \\&\_parens;

if \($string =~ m"\($\_parens\)stuff"\)
\{
\}

and have the regular expression engine&#8203;:

&nbsp;    a\) recognize that $\_parens is a code reference
&nbsp;    b\) run the code behind that coderef 
&nbsp;    c\) increment pos to the position where the first '\)' is seen\. \(as per c code\)
&nbsp;    d\) auto\-fail \(with a \-1\) if no '\)' is seen
&nbsp;    e\) continue on to match the string 'stuff' after the pos returned by the function 
&nbsp;       \_parens\.

In other words\, I want the ability to put my own C hooks inside the regex engine\.

Ed

**Comment by p5pRT** [comment]
### From @rgs

Edward Peschko wrote&#8203;:
> my $\_parens = \\&\_parens;
> 
> if \($string =~ m"\($\_parens\)stuff"\)
> \{
> \}
> 
> and have the regular expression engine&#8203;:
> 
>     a\) recognize that $\_parens is a code reference

The problem is that the regexp engine won't see the $\_parens variable\, but
something like "\(CODE\(0xdeadbeef\)\)stuff"\.

**Comment by p5pRT** [comment]
### From sky@nanisky.com


On torsdag\, maj 15\, 2003\, at 10&#8203;:12 Europe/Stockholm\, Rafael 
Garcia\-Suarez wrote&#8203;:

> The problem is that the regexp engine won't see the $\_parens variable\, 
> but
> something like "\(CODE\(0xdeadbeef\)\)stuff"\.

And this is a problem? ;\-\)

Arthur

**Comment by p5pRT** [comment]
### From esp5@mdssirds.comp.pge.com

On Thu\, May 15\, 2003 at 08&#8203;:15&#8203;:07AM \-0000\, Rafael Garcia\-Suarez wrote&#8203;:
> Edward Peschko wrote&#8203;:

> > my $\_parens = \\&\_parens;
> > 
> > if \($string =~ m"\($\_parens\)stuff"\)
> > \{
> > \}
> > 
> > and have the regular expression engine&#8203;:
> > 
> >     a\) recognize that $\_parens is a code reference
> 
> The problem is that the regexp engine won't see the $\_parens variable\, but
> something like "\(CODE\(0xdeadbeef\)\)stuff"\.

I can think of at least three ways around this problem&#8203;:

&nbsp;    1\) When a coderef stringifies\, perl remembers how to translate that coderef back 
&nbsp;       into a subroutine \(keeps a hash of coderef string to coderef value pairs\)

&nbsp;    2\) the special\, modified\, version of "" that the regex engine uses is modified such
&nbsp;       that it doesn't stringify $\_coderef\, and instead gets a callback which it calls
&nbsp;       upon entering scope\, as discussed\.

&nbsp;    3\) a special \(?\.\.\.\) syntax is put around $\_coderef to tell the regex engine that 
&nbsp;       it is dealing with a coderef\.

This is a big one \- right now\, I'm using perl to parse heavy\-duty amounts of C\+\+ code and
the regex engine is showing its strain\.  It'd be really cool to link the parts of the gcc
parser into the regex engine via Inline&#8203;::C and get a syntax\-tree 'from the horses mouth' 
so to speak\. 

Either that\, or simply get the speed benefits that C would offer in complicated text 
searching\. 

Ed

\(ps \- who is the main maintainer of the regex engine now\, anyways? Or is it unclaimed 
territory?\)

**Comment by p5pRT** [comment]
### From @pjcj

On Sat\, May 17\, 2003 at 08&#8203;:06&#8203;:16PM \-0700\, Edward Peschko wrote&#8203;:

> \(ps \- who is the main maintainer of the regex engine now\, anyways? Or
> is it unclaimed territory?\)

That would be Hugo\, I suppose\, though he would appear still to be
somewhat busy with other duties at the moment\.

\-\- 
Paul Johnson \- paul@&#8203;pjcj\.net
http&#8203;://www\.pjcj\.net

**Comment by p5pRT** [comment]
### From @cpansprout

This was fixed in perl 5\.10\.0 when the regexp engine was made iterative\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
